search/views.py

```diff
@@ -318,12 +121,109 @@ def search(request):
     else:
         return render(request, 'search/search_ajax.html', tvars)
 
 
+def _get_ids_in_cluster(request, requested_cluster_id):
+    """Get the sound ids in the requested cluster. Used for applying a filter by id when using a cluster facet.
+    """
+    try:
+        requested_cluster_id = int(requested_cluster_id) - 1
+    
+        # results are cached in clustering_utilities, available features are defined in the clustering settings file.
+        result = cluster_sound_results(request, features=DEFAULT_FEATURES)
+        results = result['result']
+
+        sounds_from_requested_cluster = results[int(requested_cluster_id)]
+
+    except ValueError:
+        return []
+    except IndexError:
+        return []
+
+    return sounds_from_requested_cluster
+
+
+def clustering_facet(request):
+    """Triggers the computation of the clustering, returns the state of processing or the clustering facet.
+    """
+    # pass the url query params for later sending it to the clustering engine
+    url_query_params_string = request.META['QUERY_STRING']
+    # remove existing cluster facet filter from the params since the returned cluster facets will include 
+    # their correspondinng cluster_id query parameter (done in the template)
+    url_query_params_string = re.sub(r"(&cluster_id=[0-9]*)", "", url_query_params_string)
+
+    result = cluster_sound_results(request, features=DEFAULT_FEATURES)
+
+    # check if computation is finished. If not, send computation state.
+    if result['finished']:
+        if result['result'] is not None:
+            results = result['result']
+            num_clusters = num_clusters = len(results) + 1
+        else:
+             return JsonResponse({'status': 'failed'}, safe=False)
+    elif result['error']:
+        return JsonResponse({'status': 'failed'}, safe=False)
+    else:
+        return JsonResponse({'status': 'pending'}, safe=False)
+
+    num_sounds_per_cluster = [len(cluster) for cluster in results]
+    classes = {sound_id: cluster_id for cluster_id, cluster in enumerate(results) for sound_id in cluster}
+
+    # label clusters using most occuring tags
+    sound_instances = sounds.models.Sound.objects.bulk_query_id(map(int, classes.keys()))
+    sound_tags = {sound.id: sound.get_sound_tags() for sound in sound_instances}
+    cluster_tags = defaultdict(list)
+    query_terms = {t.lower() for t in request.GET.get('q', '').split(' ')}
+    for sound_id, tags in sound_tags.iteritems():
+        cluster_tags[classes[str(sound_id)]] += [t.lower() for t in tags if t.lower() not in query_terms]
+    cluster_tags_with_count = {k: sorted([(t, tags.count(t)) for t in set(tags)], 
+                                         key=lambda x: x[1], reverse=True)
+                               for k, tags in cluster_tags.iteritems()}
+    cluster_most_occuring_tags = [' '.join(zip(*tags[:3])[0]) for tags in cluster_tags_with_count.values() if len(tags)>2]  # dict values sorted?!
```

Comment:
All this logic here is a bit obscure. I think you could simplify it using [`collections.Counter`](https://docs.python.org/2/library/collections.html#collections.Counter). You can pass the raw list of tags (with repeated tags) and do something like `Counter(raw_tags).most_common(3)`. This will return the 3 most common ;)